Ical.Net.Tests/MatchTimeZoneTests.cs

```diff
@@ -0,0 +1,233 @@
+﻿//
+// Copyright ical.net project maintainers and contributors.
+// Licensed under the MIT license.
+//
+
+using System;
+using System.Linq;
+using Ical.Net.DataTypes;
+using NUnit.Framework;
+
+namespace Ical.Net.Tests;
+
+[TestFixture]
+public class MatchTimeZoneTests
+{
+    [Test, Category("Recurrence")]
+    public void MatchTimeZone_LocalTimeUsaWithTimeZone()
+    {
+        // DTSTART with local time and time zone reference (negative offset), UNTIL as UTC
+        const string ical =
+            """
+            BEGIN:VCALENDAR
+            VERSION:2.0
+            PRODID:-//Example Corp//NONSGML Event//EN
+            BEGIN:VEVENT
+            UID:example1
+            SUMMARY:Event with local time and time zone
+            DTSTART;TZID=America/New_York:20231101T090000
+            RRULE:FREQ=DAILY;UNTIL=20231105T130000Z
+            DTEND;TZID=America/New_York:20231101T100000
+            END:VEVENT
+            END:VCALENDAR
+            """;
+
+        var calendar = Calendar.Load(ical);
+        var evt = calendar.Events.First();
+        var until = evt.RecurrenceRules.First().Until;
+
+        var expectedUntil = new DateTime(2023, 11, 05, 13, 00, 00, DateTimeKind.Utc);
+        var occurrences = evt.GetOccurrences(new CalDateTime(2023, 11, 01), new CalDateTime(2023, 11, 06));
+
+        Assert.Multiple(() =>
+        {
+            Assert.That(until, Is.EqualTo(expectedUntil));
+            Assert.That(occurrences.Count, Is.EqualTo(4));
+            /*
+               Should have 4 occurrences:
+               November 1, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+               November 2, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+               November 3, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+               November 4, 2023: 09:00 AM - 10:00 AM (UTC-0400) (America/New_York)
+
+               November 5, 2023: 09:00 AM - 10:00 AM (UTC-0500) (America/New_York)
+               must NOT be included, because 20231105T130000Z => November 5, 2023: 08:00 AM (America/New_York)
+               (Daylight Saving Time in America/New_York ended on Sunday, November 5, 2023, at 2:00 AM)
+           */
+        });
+    }
+
+    [Test, Category("Recurrence")]
+    public void MatchTimeZone_LocalTimeJapanWithTimeZone()
+    {
+        // DTSTART with local time and time zone reference (positive offset), UNTIL as UTC
+        const string ical =
+            """
+            BEGIN:VCALENDAR
+            VERSION:2.0
+            PRODID:-//Example Corp//NONSGML Event//EN
+            BEGIN:VEVENT
+            UID:example1
+            SUMMARY:Event with local time and time zone
+            DTSTART;TZID=Asia/Tokyo:20231101T090000
```

Comment:
In the test above with NY tz you chose dates involving a DST change (in 2023 in NY winter DST change was Nov 5), but in this test no DST change is involved (there is no DST in Tokyo). Not a problem, just noting this to make sure, its intentional.

Reply:
Good point. I updated the test to use Sydney/Australia with DST change.